File: frontend/src/lib/api/reports.ts
import { apiClient } from '../api-client';

// Downtime Report Types
export interface MachineDowntimeStats {
  machineId: number;
  machineName: string;
  departmentId: number;
  departmentName?: string;
  totalDowntime: number;
  frequency: number;
  avgDowntime: number;
}

export interface DepartmentDowntimeStats {
  departmentId: number;
  departmentName?: string;
  totalDowntime: number;
  frequency: number;
  avgDowntime: number;
}

export interface DowntimeReportResponse {
  totalDowntimeHours: number;
  totalDowntimeMinutes: number;
  frequency: number;
  avgDowntimeHours: number;
  avgDowntimeMinutes: number;
  byMachine: MachineDowntimeStats[];
  byDepartment: DepartmentDowntimeStats[];
}

export interface DowntimeReportFilters {
  machineId?: number;
  departmentId?: number;
  startDate?: string;
  endDate?: string;
  export?: 'csv' | 'excel';
}

// Maintenance Cost Report Types
export interface MachineCostStats {
  machineId: number;
  machineName: string;
  partsCost: number;
  laborCost: number;
  totalCost: number;
  maintenanceCount: number;
}

export interface MaintenanceTypeCostStats {
  maintenanceTypeId: number;
  maintenanceTypeName: string;
  partsCost: number;
  laborCost: number;
  totalCost: number;
  maintenanceCount: number;
}

export interface MaintenanceCostReportResponse {
  totalPartsCost: number;
  totalLaborCost: number;
  totalCost: number;
  byMachine: MachineCostStats[];
  byMaintenanceType: MaintenanceTypeCostStats[];
}

export interface MaintenanceCostReportFilters {
  machineId?: number;
  maintenanceTypeId?: number;
  startDate?: string;
  endDate?: string;
  export?: 'csv' | 'excel';
}

// Failure Analysis Report Types
export interface FailurePattern {
  failureCodeId: number;
  failureCode: string;
  failureDescription: string;
  failureCategory?: string;
  frequency: number;
  affectedMachineCount: number;
  affectedMachines: number[];
  avgResolutionTimeMinutes: number;
  resolutionCount: number;
}

export interface FailureAnalysisReportResponse {
  totalFailures: number;
  uniqueFailureCodes: number;
  failurePatterns: FailurePattern[];
  recurringIssues: FailurePattern[];
}

export interface FailureAnalysisReportFilters {
  machineId?: number;
  departmentId?: number;
  startDate?: string;
  endDate?: string;
  failureCategory?: string;
  export?: 'csv' | 'excel';
}

// Inventory Report Types
export interface StockLevelItem {
  id: number;
  partNumber: string;
  partName: string;
  description?: string;
  categoryNumber?: string;
  categoryName?: string;
  quantity: number;
  minQuantity: number;
  maxQuantity?: number;
  unitPrice?: number;
  location?: string;
  status: string;
}

export interface StockLevelsReportResponse {
  items: StockLevelItem[];
  totalItems: number;
  criticalCount: number;
  lowStockCount: number;
}

export interface StockLevelsFilters {
  groupNumber?: string;
  groupName?: string;
  location?: string;
  export?: 'csv' | 'excel';
}

export interface ConsumptionItem {
  partId: number;
  partNumber: string;
  partName: string;
  categoryNumber?: string;
  categoryName?: string;
  location?: string;
  quantityConsumed: number;
  totalValue: number;
}

export interface ConsumptionReportResponse {
  totalConsumption: number;
  byPart: ConsumptionItem[];
  transactionCount: number;
}

export interface ConsumptionFilters {
  dateFrom?: string;
  dateTo?: string;
  machineId?: number;
  groupNumber?: string;
  location?: string;
  export?: 'csv' | 'excel';
}

export interface ValuationByGroup {
  groupNumber: string;
  groupName: string;
  totalValuation: number;
  partCount: number;
}

export interface ValuationReportResponse {
  totalValuation: number;
  byGroup: ValuationByGroup[];
}

export interface ValuationFilters {
  groupNumber?: string;
  export?: 'csv' | 'excel';
}

export interface ReorderItem {
  id: number;
  partNumber: string;
  partName: string;
  categoryNumber?: string;
  categoryName?: string;
  location?: string;
  currentQuantity: number;
  minQuantity: number;
  shortfall: number;
  suggestedReorderQty: number;
  unitPrice?: number;
  estimatedCost: number;
}

export interface ReorderReportResponse {
  items: ReorderItem[];
  totalItems: number;
}

export interface ReorderFilters {
  groupNumber?: string;
  location?: string;
  export?: 'csv' | 'excel';
}

// Reports API
export const reportsApi = {
  // Get downtime report
  getDowntimeReport: async (filters: DowntimeReportFilters = {}): Promise<DowntimeReportResponse | Blob> => {
    const params = new URLSearchParams();
    
    if (filters.machineId) params.append('machineId', filters.machineId.toString());
    if (filters.departmentId) params.append('departmentId', filters.departmentId.toString());
    if (filters.startDate) params.append('startDate', filters.startDate);
    if (filters.endDate) params.append('endDate', filters.endDate);
    if (filters.export) params.append('export', filters.export);

    const responseType = filters.export ? 'blob' : 'json';
    const response = await apiClient.get(`/reports/downtime?${params.toString()}`, {
      responseType,
    });
    
    return response.data;
  },

  // Get maintenance cost report
  getMaintenanceCostReport: async (filters: MaintenanceCostReportFilters = {}): Promise<MaintenanceCostReportResponse | Blob> => {
    const params = new URLSearchParams();
    
    if (filters.machineId) params.append('machineId', filters.machineId.toString());
    if (filters.maintenanceTypeId) params.append('maintenanceTypeId', filters.maintenanceTypeId.toString());
    if (filters.startDate) params.append('startDate', filters.startDate);
    if (filters.endDate) params.append('endDate', filters.endDate);
    if (filters.export) params.append('export', filters.export);

    const responseType = filters.export ? 'blob' : 'json';
    const response = await apiClient.get(`/reports/maintenance-costs?${params.toString()}`, {
      responseType,
    });
    
    return response.data;
  },

  // Get failure analysis report
  getFailureAnalysisReport: async (filters: FailureAnalysisReportFilters = {}): Promise<FailureAnalysisReportResponse | Blob> => {
    const params = new URLSearchParams();
    
    if (filters.machineId) params.append('machineId', filters.machineId.toString());
    if (filters.departmentId) params.append('departmentId', filters.departmentId.toString());
    if (filters.startDate) params.append('startDate', filters.startDate);
    if (filters.endDate) params.append('endDate', filters.endDate);
    if (filters.failureCategory) params.append('failureCategory', filters.failureCategory);
    if (filters.export) params.append('export', filters.export);

    const responseType = filters.export ? 'blob' : 'json';
    const response = await apiClient.get(`/reports/failure-analysis?${params.toString()}`, {
      responseType,
    });
    
    return response.data;
  },

  // Get inventory stock levels report
  getInventoryStockLevels: async (filters: StockLevelsFilters = {}): Promise<StockLevelsReportResponse | Blob> => {
    const params = new URLSearchParams();
    
    if (filters.groupNumber) params.append('groupNumber', filters.groupNumber);
    if (filters.groupName) params.append('groupName', filters.groupName);
    if (filters.location) params.append('location', filters.location);
    if (filters.export) params.append('export', filters.export);

    const responseType = filters.export ? 'blob' : 'json';
    const response = await apiClient.get(`/reports/inventory/stock-levels?${params.toString()}`, {
      responseType,
    });
    
    return response.data;
  },

  // Get inventory consumption report
  getInventoryConsumption: async (filters: ConsumptionFilters = {}): Promise<ConsumptionReportResponse | Blob> => {
    const params = new URLSearchParams();
    
    if (filters.dateFrom) params.append('dateFrom', filters.dateFrom);
    if (filters.dateTo) params.append('dateTo', filters.dateTo);
    if (filters.machineId) params.append('machineId', filters.machineId.toString());
    if (filters.groupNumber) params.append('groupNumber', filters.groupNumber);
    if (filters.location) params.append('location', filters.location);
    if (filters.export) params.append('export', filters.export);

    const responseType = filters.export ? 'blob' : 'json';
    const response = await apiClient.get(`/reports/inventory/consumption?${params.toString()}`, {
      responseType,
    });
    
    return response.data;
  },

  // Get inventory valuation report
  getInventoryValuation: async (filters: ValuationFilters = {}): Promise<ValuationReportResponse | Blob> => {
    const params = new URLSearchParams();
    
    if (filters.groupNumber) params.append('groupNumber', filters.groupNumber);
    if (filters.export) params.append('export', filters.export);

    const responseType = filters.export ? 'blob' : 'json';
    const response = await apiClient.get(`/reports/inventory/valuation?${params.toString()}`, {
      responseType,
    });
    
    return response.data;
  },

  // Get reorder report
  getInventoryReorder: async (filters: ReorderFilters = {}): Promise<ReorderReportResponse | Blob> => {
    const params = new URLSearchParams();
    
    if (filters.groupNumber) params.append('groupNumber', filters.groupNumber);
    if (filters.location) params.append('location', filters.location);
    if (filters.export) params.append('export', filters.export);

    const responseType = filters.export ? 'blob' : 'json';
    const response = await apiClient.get(`/reports/inventory/reorder?${params.toString()}`, {
      responseType,
    });
    
    return response.data;
  },

  // Helper function to download export as file
  downloadExport: (blob: Blob, filename: string) => {
    const url = window.URL.createObjectURL(blob);
    const link = document.createElement('a');
    link.href = url;
    link.download = filename;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    window.URL.revokeObjectURL(url);
  },
};

File: backend/app/api/v1/endpoints/reports.py
"""
Reports API endpoints for maintenance analytics and reporting.
"""
from fastapi import APIRouter, Depends, Query, Request, HTTPException
from fastapi import status
from sqlalchemy.orm import Session
from typing import Optional
from datetime import datetime
import csv
import io

from app.core.database import get_db
from app.core.deps import get_current_user, require_role_list
from app.models.user import User
from app.services.maintenance_analysis_service import (
    get_downtime_statistics,
    calculate_maintenance_costs,
    analyze_failure_patterns,
    get_stock_levels,
    get_consumption_trends,
    calculate_inventory_valuation,
    get_reorder_report
)
from app.schemas.report import (
    DowntimeReportResponse,
    MaintenanceCostReportResponse,
    FailureAnalysisReportResponse,
    StockLevelsReportResponse,
    ConsumptionReportResponse,
    ValuationReportResponse,
    ReorderReportResponse,
    StockLevelItem,
    ConsumptionItem,
    ValuationByGroup,
    ReorderItem
)
from app.services.audit_service import log_activity

router = APIRouter()

# Roles that can access reports: Admin and Maintenance Manager
REPORT_ALLOWED_ROLES = ["ADMIN", "MAINTENANCE_MANAGER"]

# Roles that can access inventory reports: Admin and Inventory Manager
INVENTORY_REPORT_ROLES = ["ADMIN", "INVENTORY_MANAGER"]


@router.get("/downtime", response_model=DowntimeReportResponse)
async def get_downtime_report(
    machineId: Optional[int] = Query(None, description="Filter by machine ID"),
    departmentId: Optional[int] = Query(None, description="Filter by department ID"),
    startDate: Optional[datetime] = Query(None, description="Filter from date"),
    endDate: Optional[datetime] = Query(None, description="Filter to date"),
    export: Optional[str] = Query(None, description="Export format: csv or excel"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
    request: Request = None
):
    """
    Get downtime statistics report.
    
    Admin and Maintenance Manager access only.
    """
    # Check if user has required role
    if current_user.role not in REPORT_ALLOWED_ROLES:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Insufficient permissions"
        )
    
    # Get downtime statistics
    stats = get_downtime_statistics(
        db=db,
        machine_id=machineId,
        department_id=departmentId,
        start_date=startDate,
        end_date=endDate
    )
    
    # Log activity
    log_activity(
        db=db,
        userId=current_user.id,
        action="READ",
        entityType="DOWNTIME_REPORT",
        entityId=0,
        description=f"Accessed downtime report",
        request=request
    )
    db.commit()
    
    # Check if export requested
    if export and export.lower() == 'csv':
        return await _export_downtime_csv(stats, startDate, endDate)
    
    return stats


@router.get("/maintenance-costs", response_model=MaintenanceCostReportResponse)
async def get_maintenance_cost_report(
    machineId: Optional[int] = Query(None, description="Filter by machine ID"),
    maintenanceTypeId: Optional[int] = Query(None, description="Filter by maintenance type ID"),
    startDate: Optional[datetime] = Query(None, description="Filter from date"),
    endDate: Optional[datetime] = Query(None, description="Filter to date"),
    export: Optional[str] = Query(None, description="Export format: csv or excel"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
    request: Request = None
):
    """
    Get maintenance cost analysis report.
    
    Admin and Maintenance Manager access only.
    """
    # Check if user has required role
    if current_user.role not in REPORT_ALLOWED_ROLES:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Insufficient permissions"
        )
    
    # Get cost analysis
    costs = calculate_maintenance_costs(
        db=db,
        machine_id=machineId,
        maintenance_type_id=maintenanceTypeId,
        start_date=startDate,
        end_date=endDate
    )
    
    # Log activity
    log_activity(
        db=db,
        userId=current_user.id,
        action="READ",
        entityType="COST_REPORT",
        entityId=0,
        description=f"Accessed maintenance cost report",
        request=request
    )
    db.commit()
    
    # Check if export requested
    if export and export.lower() == 'csv':
        return await _export_costs_csv(costs, startDate, endDate)
    
    return costs


@router.get("/failure-analysis", response_model=FailureAnalysisReportResponse)
async def get_failure_analysis_report(
    machineId: Optional[int] = Query(None, description="Filter by machine ID"),
    departmentId: Optional[int] = Query(None, description="Filter by department ID"),
    startDate: Optional[datetime] = Query(None, description="Filter from date"),
    endDate: Optional[datetime] = Query(None, description="Filter to date"),
    failureCategory: Optional[str] = Query(None, description="Filter by failure category"),
    export: Optional[str] = Query(None, description="Export format: csv or excel"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
    request: Request = None
):
    """
    Get failure analysis report with pattern identification.
    
    Admin and Maintenance Manager access only.
    """
    # Check if user has required role
    if current_user.role not in REPORT_ALLOWED_ROLES:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Insufficient permissions"
        )
    
    # Get failure analysis
    analysis = analyze_failure_patterns(
        db=db,
        machine_id=machineId,
        department_id=departmentId,
        start_date=startDate,
        end_date=endDate,
        failure_category=failureCategory
    )
    
    # Log activity
    log_activity(
        db=db,
        userId=current_user.id,
        action="READ",
        entityType="FAILURE_REPORT",
        entityId=0,
        description=f"Accessed failure analysis report",
        request=request
    )
    db.commit()
    
    # Check if export requested
    if export and export.lower() == 'csv':
        return await _export_failure_csv(analysis, startDate, endDate)
    
    return analysis


async def _export_downtime_csv(stats: dict, start_date: Optional[datetime], end_date: Optional[datetime]):
    """Export downtime report to CSV."""
    from fastapi.responses import StreamingResponse
    
    output = io.StringIO()
    writer = csv.writer(output)
    
    # Write header
    writer.writerow([
        "Report Type: Downtime Statistics",
        "Date Range: {} to {}".format(
            start_date.strftime('%Y-%m-%d') if start_date else "All",
            end_date.strftime('%Y-%m-%d') if end_date else "All"
        )
    ])
    writer.writerow([])
    
    # Write summary
    writer.writerow(["Summary"])
    writer.writerow(["Total Downtime (Hours)", stats['totalDowntimeHours']])
    writer.writerow(["Total Downtime (Minutes)", stats['totalDowntimeMinutes']])
    writer.writerow(["Frequency", stats['frequency']])
    writer.writerow(["Average Downtime (Hours)", stats['avgDowntimeHours']])
    writer.writerow(["Average Downtime (Minutes)", stats['avgDowntimeMinutes']])
    writer.writerow([])
    
    # Write by machine
    writer.writerow(["Downtime by Machine"])
    writer.writerow(["Machine ID", "Machine Name", "Department", "Total Downtime (Hours)", "Frequency", "Avg Downtime (Hours)"])
    for machine in stats['byMachine']:
        writer.writerow([
            machine['machineId'],
            machine['machineName'],
            machine['departmentName'] or '',
            machine['totalDowntime'],
            machine['frequency'],
            machine['avgDowntime']
        ])
    writer.writerow([])
    
    # Write by department
    writer.writerow(["Downtime by Department"])
    writer.writerow(["Department ID", "Department Name", "Total Downtime (Hours)", "Frequency", "Avg Downtime (Hours)"])
    for dept in stats['byDepartment']:
        writer.writerow([
            dept['departmentId'],
            dept['departmentName'] or '',
            dept['totalDowntime'],
            dept['frequency'],
            dept['avgDowntime']
        ])
    
    output.seek(0)
    filename = f"downtime_report_{datetime.utcnow().strftime('%Y%m%d_%H%M%S')}.csv"
    
    return StreamingResponse(
        iter([output.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": f"attachment; filename={filename}"}
    )


async def _export_costs_csv(costs: dict, start_date: Optional[datetime], end_date: Optional[datetime]):
    """Export maintenance cost report to CSV."""
    from fastapi.responses import StreamingResponse
    
    output = io.StringIO()
    writer = csv.writer(output)
    
    # Write header
    writer.writerow([
        "Report Type: Maintenance Cost Analysis",
        "Date Range: {} to {}".format(
            start_date.strftime('%Y-%m-%d') if start_date else "All",
            end_date.strftime('%Y-%m-%d') if end_date else "All"
        )
    ])
    writer.writerow([])
    
    # Write summary
    writer.writerow(["Summary"])
    writer.writerow(["Total Parts Cost", costs['totalPartsCost']])
    writer.writerow(["Total Labor Cost", costs['totalLaborCost']])
    writer.writerow(["Total Cost", costs['totalCost']])
    writer.writerow([])
    
    # Write by machine
    writer.writerow(["Costs by Machine"])
    writer.writerow(["Machine ID", "Machine Name", "Parts Cost", "Labor Cost", "Total Cost", "Maintenance Count"])
    for machine in costs['byMachine']:
        writer.writerow([
            machine['machineId'],
            machine['machineName'],
            machine['partsCost'],
            machine['laborCost'],
            machine['totalCost'],
            machine['maintenanceCount']
        ])
    writer.writerow([])
    
    # Write by maintenance type
    writer.writerow(["Costs by Maintenance Type"])
    writer.writerow(["Maintenance Type ID", "Type Name", "Parts Cost", "Labor Cost", "Total Cost", "Maintenance Count"])
    for maint_type in costs['byMaintenanceType']:
        writer.writerow([
            maint_type['maintenanceTypeId'],
            maint_type['maintenanceTypeName'],
            maint_type['partsCost'],
            maint_type['laborCost'],
            maint_type['totalCost'],
            maint_type['maintenanceCount']
        ])
    
    output.seek(0)
    filename = f"maintenance_costs_report_{datetime.utcnow().strftime('%Y%m%d_%H%M%S')}.csv"
    
    return StreamingResponse(
        iter([output.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": f"attachment; filename={filename}"}
    )


async def _export_failure_csv(analysis: dict, start_date: Optional[datetime], end_date: Optional[datetime]):
    """Export failure analysis report to CSV."""
    from fastapi.responses import StreamingResponse
    
    output = io.StringIO()
    writer = csv.writer(output)
    
    # Write header
    writer.writerow([
        "Report Type: Failure Analysis",
        "Date Range: {} to {}".format(
            start_date.strftime('%Y-%m-%d') if start_date else "All",
            end_date.strftime('%Y-%m-%d') if end_date else "All"
        )
    ])
    writer.writerow([])
    
    # Write summary
    writer.writerow(["Summary"])
    writer.writerow(["Total Failures", analysis['totalFailures']])
    writer.writerow(["Unique Failure Codes", analysis['uniqueFailureCodes']])
    writer.writerow([])
    
    # Write failure patterns
    writer.writerow(["Failure Patterns"])
    writer.writerow([
        "Failure Code ID",
        "Code",
        "Description",
        "Category",
        "Frequency",
        "Affected Machines",
        "Avg Resolution Time (Minutes)",
        "Resolution Count"
    ])
    for pattern in analysis['failurePatterns']:
        writer.writerow([
            pattern['failureCodeId'],
            pattern['failureCode'],
            pattern['failureDescription'],
            pattern['failureCategory'] or '',
            pattern['frequency'],
            pattern['affectedMachineCount'],
            pattern['avgResolutionTimeMinutes'],
            pattern['resolutionCount']
        ])
    writer.writerow([])
    
    # Write recurring issues
    writer.writerow(["Recurring Issues (Top 10)"])
    writer.writerow([
        "Failure Code ID",
        "Code",
        "Description",
        "Category",
        "Frequency",
        "Affected Machines",
        "Avg Resolution Time (Minutes)"
    ])
    for issue in analysis['recurringIssues']:
        writer.writerow([
            issue['failureCodeId'],
            issue['failureCode'],
            issue['failureDescription'],
            issue['failureCategory'] or '',
            issue['frequency'],
            issue['affectedMachineCount'],
            issue['avgResolutionTimeMinutes']
        ])
    
    output.seek(0)
    filename = f"failure_analysis_report_{datetime.utcnow().strftime('%Y%m%d_%H%M%S')}.csv"
    
    return StreamingResponse(
        iter([output.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": f"attachment; filename={filename}"}
    )

